File: 100.same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    # Approach: BFS, Complexity: O(m+n), O(m+n)
    # Tip: There a one-one relation between a tree and a level order traversal.

    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        def bfs(root: Optional[TreeNode]) -> List[int|str]:
            traversal = []
            q = deque()
            q.append(root)

            while q:
                node = q.popleft()
                if not node:
                    traversal.append('-')
                    continue

                traversal.append(node.val)
                q.append(node.left)
                q.append(node.right)

            return traversal

        return bfs(p) == bfs(q)

File: 100.same-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

/* Approach: DFS + Recursion, Complexity: O(n), O(n) */
/* Tip: Two trees are the same if all of their corresponing subtrees are same. */

bool isSameTree(struct TreeNode* p, struct TreeNode* q){
    /* Paths in same trees terminate simultaneously and nodes have equal values. */
    if (!p && !q)
        return true;
    else if (!p ^ !q || p->val != q->val)
        return false;

    /* Corresponding left and right subtrees must be same. */
    return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
}
